app/controls.go

```diff
@@ -0,0 +1,106 @@
+package main
+
+import (
+	"log"
+	"math/rand"
+	"net/http"
+	"net/rpc"
+	"sync"
+
+	"github.com/gorilla/mux"
+
+	"github.com/weaveworks/scope/xfer"
+)
+
+const (
+	controlKey = "control"
+	originKey  = "originNodeId"
+)
+
+// ControlRouter handles the impedence mismatch between posted Control calls,
+// and websocket connections.
+type ControlRouter interface {
+	HandleControl(w http.ResponseWriter, r *http.Request)
+	HandleProbeWS(w http.ResponseWriter, r *http.Request)
+	Stop()
+}
+
+type controlRouter struct {
+	sync.Mutex
+	quit   chan struct{}
+	probes map[string]xfer.ControlHandler
+}
+
+// NewControlRouter makes a new ControlRouter
+func NewControlRouter() ControlRouter {
+	return &controlRouter{
+		probes: map[string]xfer.ControlHandler{},
+	}
+}
+
+func (cr *controlRouter) Stop() {
+	close(cr.quit)
+}
+
+// HandleControl ...
+func (cr *controlRouter) HandleControl(w http.ResponseWriter, r *http.Request) {
+	var (
+		vars    = mux.Vars(r)
+		probeID = vars["probeID"]
+		nodeID  = vars["nodeID"]
+		control = vars["control"]
+	)
+	handler, ok := cr.probes[probeID]
+	if !ok {
+		log.Printf("Probe %s is not connected right now...", probeID)
+		http.NotFound(w, r)
+		return
+	}
+
+	result := handler(xfer.Request{
+		ID:      rand.Int63(),
+		NodeID:  nodeID,
+		Control: control,
+	})
+	if result.Error != "" {
+		respondWith(w, http.StatusBadRequest, result.Error)
+		return
+	}
+	respondWith(w, http.StatusOK, result.Value)
+}
+
+func (cr *controlRouter) HandleProbeWS(w http.ResponseWriter, r *http.Request) {
+	probeid := r.Header.Get(xfer.ScopeProbeIDHeader)
+	if probeid == "" {
+		respondWith(w, http.StatusBadRequest, xfer.ScopeProbeIDHeader)
+		return
+	}
+
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Printf("Error upgrading to websocket: %v", err)
+		return
+	}
+	defer conn.Close()
+
+	codec := xfer.NewJSONWebsocketCodec(conn)
+	client := rpc.NewClientWithCodec(codec)
+
+	cr.Lock()
+	cr.probes[probeid] = func(req xfer.Request) xfer.Response {
+		var res xfer.Response
+		client.Call("control.Handle", req, &res)
+		return res
+	}
+	cr.Unlock()
+
+	defer func() {
```

Comment:
Actually there is a race here; if a probe reconnects before the close notify drops the last connection, we'll end up deleting the new connection from the map, not the old one.